feat: implement 0x20 - SHA3 Opcode

Quentash opened the pull request:
Implementation of SHA3 opcode (0x20).
The code has been run against official ethereum tests found at https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/VMTests/vmTests/sha3Filler.yml#L52

Only a few tests couldn't be run because the actual `memory.bytes_len` is a u32 and some tested value are u64, It has been discussed with @Eikix to leave it as it is for now and eventually refactor it later to handle u64 memory size.

I have added comments in the code ( to eventually remove later ) to explain how the code bypasses continuous memory allocation when trying to access a memory slot beyond `memory.bytes_len` and will instead ensure that the memory is allocated at the end of the process, this seems way more efficient since some tests will fail otherwise.

## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Resolves: #115

## What is the new behavior?

<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

<!-- If this does introduce a breaking change, please describe the impact and migration path for existing applications below. -->

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,10 +1,103 @@
+        let mut size: u32 = Into::<u256, Result<u32, EVMError>>::into((self.stack.pop()?))?;
+        let init_size = size;
+
+        let mut toHash: Array<u64> = ArrayTrait::<u64>::new();
```
```suggestion
        let mut to_hash: Array<u64> = Default::default();
```

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,10 +1,103 @@
+                break;
+            }
+            // If we try to read unallocted memory slot, we'll feed the data to hash with 0s
+            // and allocated the memory space the end of the process, which is cheaper than allocating every load().
```
```suggestion
            // and allocate the memory space at the end of the process, which is cheaper than allocating every load().
```

is this what you meant?

Quentash replied:
It is, indeed, mb

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,10 +1,103 @@
+            if size < 32 {
+                break;
+            }
+            // If we try to read unallocted memory slot, we'll feed the data to hash with 0s
```
```suggestion
            // If we try to read unallocted memory slot, we'll pad left the data to hash with 0s
```

is this what you meant?

enitrat replied:
A comment that reflects the actions performed in the code is not useful - the comment must inform on the why, not the how. Here if I understand well, it would be something along the lines of

```rust
// Pad left the hash input with zeroes for out-of-bound bytes
```

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,10 +1,103 @@
+            loaded.high = integer::u128_byte_reverse(loaded.high);
+
+            // Split the loaded word into u64 to feed cairo_keccak
+            let (highL, lowL) = u128_split(loaded.high);
```
snake_case everywhere

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,10 +1,103 @@
+            size -= 32;
+        };
+
+        if size > 0 {
```
can you add a small comment here :)

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,10 +1,103 @@
+            }
+            // If we try to read unallocted memory slot, we'll feed the data to hash with 0s
+            // and allocated the memory space the end of the process, which is cheaper than allocating every load().
+            if (offset + (32 * counter)) > self.memory.bytes_len {
```
can you make `offset + (32 * counter)` a variable with an explicit name that indicates what it represents?

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,10 +1,103 @@
+
+#[generate_trait]
+impl Sha3Impl of Sha3Trait {
+    /// SHA3 operation.
```
A rustdoc-like description of what the function does would be helpful for other developers reading this. 
You need to describe the algorithm applied in natural language (e.g. context on why we apply, why we split u128s into u64s, etc

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,10 +1,103 @@
+
+        if size > 0 {
+            let mut loaded = 0;
+            if (offset + (32 * counter)) > self.memory.bytes_len {
```
```rust 
            let mut loaded = if (offset + (32 * counter)) > self.memory.bytes_len { ...
```

Quentash replied:
That is very instructive, I never realized this was doable, thanks!

enitrat commented on crates/utils/src/helpers.cairo:
```diff
@@ -102,6 +102,20 @@ fn split_word_128(value: u256, ref dst: Array<u8>) {
 }
 
+/// Try to convert u128 to u64,
+fn u128_to_u64(input: u128) -> u64 {
```
this is useless, why not simply `input.try_into().unwrap()` where it's needed?

enitrat commented on crates/utils/src/helpers.cairo:
```diff
@@ -102,6 +102,20 @@ fn split_word_128(value: u256, ref dst: Array<u8>) {
+}
+
+/// Splits a u128 into two u64.
+fn u128_split(input: u128) -> (u64, u64) {
```
I see that it's something from `corelib/src.keccak`. You can import it into your sha3 file with

`use keccak::u128_split;`

Quentash commented:
Used Default::default() as array initiator.
Corrected various comment and added some.
Added SHA3 rustdoc-like comment. ( Please tell me if this one is explicit enough ).
Added `slot_to_read` variable which made me realize I could dispose the `counter` one.
Replaced sneaky camels by snakes.

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,10 +1,110 @@
+        let mut size: u32 = Into::<u256, Result<u32, EVMError>>::into((self.stack.pop()?))?;
+        let init_size = size;
+
+        let mut to_hash: Array<u64> = Default::default();
```
this part should be split into multiple sub functions. Define internal functions inside an `internal` mod for everything related to
- Padding of input
- Transformation to u64 words
- Bytes reversing

This entire function should not be more than 10 - 20 lines long. Separate it by concern as much as you can to make the code easier to read, understand and debug. Use helper functions where required. These functions will need to be unit tested as well

Quentash replied:
I will do the same thing with the LOG opcode because it has been written the same way than this.

Quentash commented:
I have pushed a new version of the sha3 algorithm hoping it is clearer and better structured. ( There is a significant improvement in gas cost concerning big sized test 6993735300 -> 1703355056 for the bigger one. )
I still have tests to add but I pushed this commit to allow you to check on `Memory.load()`.

Indeed, it appears that the memory expansion gas cost calculation makes one of my test fail with what firstly appears to be an infinite loop that ends up with a `error: process did not exit successfully: signal: 9 (SIGKILL)` message.
If I comment those lines, my tests appear to run normally. "Funny" thing to note, this is not even the "biggest" test that fails, see screenshots.

With comments :
![image](https://github.com/kkrt-labs/kakarot-ssj/assets/100387965/38570bfa-a99a-4f17-82b3-b9d90c254952)

Without comments :
![image](https://github.com/kkrt-labs/kakarot-ssj/assets/100387965/ef7187b8-e4e9-4f6b-8cf9-173f7310eba3)

enitrat reviewed (commented):
There is still some refactoring to be done on how the function is constructed internally, otherwise it's a good work!

Here's an example of a keccak256 function taking an array of bytes as an input. As you can see it's around 12 lines long, and the intent by each line can be directly understood, even without having functions implementations available. This is ideally what you should aim for when writing code

```rust
fn cairo_keccak_le_bytes(mut self: Array<u8>) -> u256 {
    let mut words64: Array<u64> = array![];
    let mut self = self.span();
    let (last_word, last_word_bytes) = loop {
        // Specifically handle last word
        if self.len() < 8 {
            let (value, n_bytes) = U64Trait::from_le_bytes(self);
            break (value, n_bytes);
        };
        let mut current_word = self.slice(0, 8);
        let (value, n_bytes) = U64Trait::from_le_bytes(current_word);
        words64.append(value);
        self = self.slice(8, self.len() - 8);
    };
    let mut hash = cairo_keccak(ref words64, last_word, last_word_bytes);
    reverse_endianness(hash)
}


fn reverse_endianness(value: u256) -> u256 {
    let new_low = integer::u128_byte_reverse(value.high);
    let new_high = integer::u128_byte_reverse(value.low);
    u256 { low: new_low, high: new_high }
}
```

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,110 @@
+    /// Return the last u64 chunk to hash, given a size, from an u256.
+    /// This function is used to prepare inputs for keccak::cairo_keccak.
+    ///
+    /// This function will split a given u256 into little endian u64 and
```
Even the (GPT-generated?) description is not accurate 😛

Quentash replied:
Hahaha !! No, I sadly don't need AI to sound confusing ! My bad !
I do understand why it is inaccurate seeing the other comments tho, will change that.

enitrat replied:
haha I said that because I noticed a phrasing pattern I often see when I ask it to generate doc 😆

Quentash commented:
> > I have pushed a new version of the sha3 algorithm hoping it is clearer and better structured. ( There is a significant improvement in gas cost concerning big sized test 6993735300 -> 1703355056 for the bigger one. ) I still have tests to add but I pushed this commit to allow you to check on `Memory.load()`.
> > Indeed, it appears that the memory expansion gas cost calculation makes one of my test fail with what firstly appears to be an infinite loop that ends up with a `error: process did not exit successfully: signal: 9 (SIGKILL)` message. If I comment those lines, my tests appear to run normally. "Funny" thing to note, this is not even the "biggest" test that fails, see screenshots.
> > With comments : ![image](https://user-images.githubusercontent.com/100387965/267426552-38570bfa-a99a-4f17-82b3-b9d90c254952.png)
> > Without comments : ![image](https://user-images.githubusercontent.com/100387965/267426441-ef7187b8-e4e9-4f6b-8cf9-173f7310eba3.png)
> 
> I tried locally and I didn't have this problem!

Ok that is even more weird. I tried it again and it gives me the same results.

Here is my scarb --version output : 
scarb 0.7.0 (58cc88efb 2023-08-23)
cairo: 2.2.0 (https://crates.io/crates/cairo-lang-compiler/2.2.0)
sierra: 1.3.0
In case yours differ ?

Otherwise, could it be because I'm on WSL ?
I finish pushing the required corrections and I will try to run the tests on a native Linux.

Would you have any other ideas ?

enitrat commented:
> Otherwise, could it be because I'm on WSL ?

There's a chance it's because of this. I'm running on macOS

Quentash commented:
Added suggested functions.
Revamped the already existing functions and comments.
Corrected typos.
Noticed a small gas improvement on big sized test, ~1%ish, but still!

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -43,53 +99,45 @@ impl Sha3Impl of Sha3Trait {
+    /// * `self` - The context in which the memory is read
+    /// * `to_hash` - A reference to the array to fill
+    /// * `amount` - The amount of 32 Bytes chunks of zeroes to append to to hash
+    fn fill_array_with_zeroes(
```
you can use `append_n` from `utils::helpers`

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -7,32 +7,88 @@ use evm::memory::MemoryTrait;
+            size, offset, self.memory.bytes_len
+        );
+        internal::fill_array_with_memory_chunks(ref self, ref to_hash, ref offset, chunks_from_mem);
+        internal::fill_array_with_zeroes(ref self, ref to_hash, chunks_of_zeroes);
```
```suggestion
    to_hash.append_n(0, chunks_of_zeroes);
```

Quentash replied:
Feeling so dumb, sorry

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,157 @@
+
+        let mut to_hash: Array<u64> = Default::default();
+
+        let (chunks_from_mem, chunks_of_zeroes) = internal::compute_data_origin(
```
```suggestion
        let (nb_chunks, nb_zeroes) = internal::compute_data_origin(
```

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,157 @@
+    /// * `offset` - Offset in memory
+    /// * `mem_len` - Size of the memory
+    /// Returns : (chunk_from_mem, chunks_of_zeroes)
+    fn compute_data_origin(size: u32, offset: u32, mem_len: u32) -> (u32, u32) {
```
@ClementWalter I don't really like this naming but I'm out of inspiration, what would you name it?

Quentash replied:
Shall I suggest "compute_memory_chunks_amount" ?

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,157 @@
+    /// * `to_hash` - A reference to the array to fill
+    /// * `offset` - Offset in memory
+    /// * `amount` - The amount of 32 Bytes chunks to read from memory
+    fn fill_array_with_memory_chunks(
```
instead of passing `offset` by ref, return it explicitly at the end of the function

enitrat commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,157 @@
+    /// * `value` - The 32 Bytes chunk to split and get the bytes from
+    /// * `size` - The amount of bytes still required to hash
+    /// Returns the last u64 chunk that isn't 8 Bytes long.
+    fn fill_array_with_last_inputs(ref to_hash: Array<u64>, value: u256, size: u32) -> u64 {
```
```suggestion
    fn prepare_last_input(ref to_hash: Array<u64>, value: u256, size: u32) -> u64 {
```

if you call it `fill_array_with_last_inputs` then an unaware reader would assume that to_hash is filled with last_inputs, where in reality, not only, we prepare the last_input arguments for the call to `cairo_keccak`

enitrat reviewed (commented):
Almost there!

Quentash commented:
I pushed a correction which include the modification of `compute_data_origin` into `compute_memory_chunks_amount` but I am totally open to any suggestions!

enitrat commented on crates/evm/src/tests/test_instructions/test_sha3.cairo:
```diff
@@ -0,0 +1,376 @@
+
+#[test]
+#[available_gas(20000000)]
+fn test_sha3_size_0_offset_0() {
```
all tests testing the `exec_sha3` function should be called `test_exec_sha3...`

This makes it easier if we want to execute only specific tests using filtering.

enitrat commented on crates/evm/src/tests/test_instructions/test_sha3.cairo:
```diff
@@ -0,0 +1,376 @@
+
+#[test]
+#[available_gas(20000000000)]
+fn test_sha3_internal_fill_array_with_memory_chunks() {
```
unit tests of internal functions should not be called `test_sha3` as they're not testing sha3

enitrat commented on crates/evm/src/tests/test_instructions/test_sha3.cairo:
```diff
@@ -0,0 +1,376 @@
+    ctx.stack.push(0x0C80);
+    ctx.stack.push(0x00);
+
+    let mut memDst: u32 = 0;
```
snake_case everywhere

enitrat commented:
@Eikix good for a final review

Eikix commented on crates/utils/src/helpers.cairo:
```diff
@@ -103,6 +104,21 @@ fn split_word_128(value: u256, ref dst: Array<u8>) {
 
+/// Splits an u256 into 4 little endian u64.
+/// Returns ((high_high, high_low),(low_high, low_low))
+fn split_u256_into_u64_little(value: u256) -> ((u64, u64), (u64, u64)) {
```
imo these two helpers should be methods implemented on the u256 type, e.g. with an auto generated trait

Eikix replied:
i also propose to have ubiquitous naming for little endian and big endian:
a `le` suffix and `be` suffix on function names, to match rust and python implementation of endianness utils

enitrat replied:
it also matches Cairo's

Quentash replied:
Agree with the proposed solutions.
I used this naming because other helpers functions already used this naming, maybe those should be changed too (ideally in an a separated PR) for the sake of consistency ? I can take care of that if you wish.

enitrat replied:
OK to do it in another PR

Eikix commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,138 @@
+    /// * `offset` - Offset in memory
+    /// * `mem_len` - Size of the memory
+    /// Returns : (nb_chunks, nb_zeroes)
+    fn compute_memory_chunks_amount(size: u32, offset: u32, mem_len: u32) -> (u32, u32) {
```
this is a bit misleading because afaik, chunks in memory usual language designates a 16-bit interval

Eikix replied:
i propose `word` to designate a 32-byte piece of data
memory `chunk` to designate a 16-byte piece of data

Eikix commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,138 @@
+        if offset > mem_len {
+            return (0, size / 32);
+        }
+        if (mem_len - offset < 32) && (size > 32) {
```
can add comments to explain what is happening?

Eikix commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,138 @@
+        if (mem_len - offset < 32) && (size > 32) {
+            return (1, (size / 32) - 1);
+        }
+        let nb_chunks = cmp::min(mem_len - offset, size) / 32;
```
can add comments?

Eikix replied:
what happens here if mem_len == 32, offset == 0 and size == 31?

Eikix commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,138 @@
+    /// * `offset` - Offset in memory to start reading from
+    /// * `amount` - The amount of 32 bytes chunks to read from memory
+    /// Return the new offset
+    fn fill_array_with_memory_chunks(
```
same remark as above

Eikix commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,138 @@
+            return (1, (size / 32) - 1);
+        }
+        let nb_chunks = cmp::min(mem_len - offset, size) / 32;
+        (nb_chunks, (size / 32) - nb_chunks)
```
what happens here if mem_len == 32, offset == 0 and size == 31? 

wouldn't (size / 32) - nb_chunks underflow?

enitrat replied:
@Quentash did you forget to push? you marked as resolve but no changes

Eikix commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,146 @@
+        let mut last_input_offset = internal::fill_array_with_memory_words(
+            ref self, ref to_hash, offset, nb_words
+        );
+        to_hash.append_n(0, 4 * nb_zeroes);
```
can you add a comment why we need to do this? 
with links to the EVM resource that explains this zeroes padding problem

Quentash replied:
I don't think this really is a problem, the idea is to append 0s this way because it is faster/cheaper than reading them from memory. 
I've added a comment which explains this tho !

Have I misunderstood/miss something ?

Eikix replied:
no problem! I was simply asking for comments explaining the choice for readability:)! no problem

Eikix commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -1,9 +1,146 @@
+        to_hash.append_n(0, 4 * nb_zeroes);
+
+        // Fill last_input with last bytes to hash
+        let last_input: u64 = if (size % 32 != 0) {
```
can you add a comment to explain the logic behind last_input manipulation?

Eikix reviewed (dismissed):
as far as i'm concerned this PR is ready
did we make sure:
- we have all the tests that geth uses?
- all the tests that EELS uses?

truly nice display of patience and good cooperation, thank you for your hard work @Quentash @enitrat

Quentash commented:
> as far as i'm concerned this PR is ready did we make sure:
> 
> * we have all the tests that geth uses?
> * all the tests that EELS uses?
> 
> truly nice display of patience and good cooperation, thank you for your hard work @Quentash @enitrat

I have taken tests found here : https://github.com/ethereum/go-ethereum/blob/master/tests/solidity/test/opCodes.js
which are quite similiar to those found here : https://github.com/ethereum/legacytests/tree/97de5bce3a257a4b219666134df15b6d4d47f1f5/Constantinople/VMTests/vmSha3Test

Beside tests with size or offset > u64.maxvalue, I've taken all I could, some are even e bit repetitve and I'm not sure they are really relevant in our case but I still added them.
I'm open to suggestions !

Eikix commented on crates/evm/src/instructions/sha3.cairo:
```diff
@@ -33,9 +33,13 @@ impl Sha3Impl of Sha3Trait {
         let mut last_input_offset = internal::fill_array_with_memory_words(
             ref self, ref to_hash, offset, nb_words
         );
+        // Fill array to hash with zeroes for bytes out of memory bound
```
do we still perform memory expansion? for bytes out of memory bound? 
do we need to?

Quentash replied:
I've just realised that we don't do the "last memory expansion" for multiple of 32 size.
But otherwise, yes, my idea was to do the memory expansion only once at the end.
Should I modify it tho ? I'm afraid I don't really understand what would be the actual impact.
Because since we only read from memory, all the "expanded" bytes would be zeroes, is it therefore relevant to update de memory length since all it contains are 0 ?

Eikix replied:
good question! I'm not sure here, as the tradeoff here is gas/performance vs. sticking to official spec
i imagine we need to perform all the expansion that the spec requires
@enitrat

enitrat replied:
Ah, yes, it used to be here at the beginning but is no longer there after several refactors. 

The thing is, ideally, memory expansions is only handled from the `memory` module, not from outside, as it's an internal mechanism to the memory. However here as we're looking for optimisations and we avoid reading zero words from memory, we might better just call `memory.ensure_length(offset+size)` directly in sha3

Quentash replied:
Roger that !

Quentash commented:
Ensure memory length to be set even if we skipped reading for zeroes.

enitrat commented:
@Quentash can you add a test that checks the memory length when we read out of bound bytes? To make sure that the length is correctly updated.

Quentash commented:
> @Quentash can you add a test that checks the memory length when we read out of bound bytes? To make sure that the length is correctly updated.

Of course, my bad !!

Quentash commented:
Added memory length verification to tests.
Made me realise it was better to ensure_length outside of the "if", thank you !

Eikix reviewed (approved):
the long awaited moment... 🔥🔥🔥🔥🔥

enitrat commented:
🥳
